### PR TITLE
feat: support in-band window resize notifications (mode 2048)

### LINKED ISSUE #2009: feat: support in-band window resize notifications (mode 2048)
## Summary

Add support for [DEC private mode 2048](https://gist.github.com/rockorager/e695fb2924d36b2bcf1fff4a3704bd83) — in-band window resize notifications. This mode places resize events into the terminal's data stream, eliminating the race conditions inherent in SIGWINCH-based resize detection.

## Motivation

Terminal resize is currently the only event that arrives out-of-band via a signal (SIGWINCH) rather than through the input stream. This creates several problems:

- **Race conditions** — the kernel delivers SIGWINCH from a different thread, on its own schedule. The application must synchronize the signal with its own state, which is inherently racy.
- **Transport limitations** — SIGWINCH doesn't propagate reliably over nested SSH sessions, some terminal multiplexers, or custom transports. This is directly relevant to JLine's `remote-ssh` and `remote-telnet` modules.
- **No pixel dimensions** — SIGWINCH carries no data; the application must call `ioctl(TIOCGWINSZ)` separately, introducing another race window. Mode 2048 reports both character and pixel dimensions inline.

## Specification

Mode 2048 uses standard DECSET/DECRST mechanics:

| Action | Sequence |
|--------|----------|
| Enable | `CSI ? 2048 h` |
| Disable | `CSI ? 2048 l` |
| Query (DECRQM) | `CSI ? 2048 $ p` |
| Report (terminal → app) | `CSI 48 t` with rows, cols, pixel width, pixel height |

Key semantics from the spec:
- When first enabled, the terminal **must** immediately report the current size
- Re-enabling while already enabled **must** also trigger an immediate report
- The terminal **must not** send notifications until the internal resize is complete
- If the terminal cannot report pixel sizes, it must report them as `0`

## Terminal support

Already shipping in 4 terminals:

| Terminal | Status |
|----------|--------|
| Ghostty | ✅ |
| foot | ✅ |
| iTerm2 | ✅ |
| kitty | ✅ |
| WezTerm | Requested (wezterm/wezterm#5826) |
| Windows Terminal | Requested (microsoft/terminal#19618) |

## Suggested approach

1. **Probe support** — extend the existing DECRQM probing infrastructure (used for mode 2027) to also probe mode 2048
2. **Terminal API** — expose enable/disable methods on `Terminal`, following the pattern of `trackFocus()` / `hasFocusSupport()`
3. **Input parsing** — recognize `CSI 48 t` reports in the input stream and translate them into resize events, as an alternative to the SIGWINCH path
4. **Remote terminals** — wire this into `remote-ssh` and `remote-telnet` where SIGWINCH forwarding is weakest and the benefit is greatest

## References

- [Mode 2048 spec](https://gist.github.com/rockorager/e695fb2924d36b2bcf1fff4a3704bd83) (rockorager)
- [Tau's DEC modes list](https://tau.garden/dec-modes/)
- [kitty #7642](https://github.com/kovidgoyal/kitty/issues/7642)
- [WezTerm #5826](https://github.com/wezterm/wezterm/issues/5826)
- [Windows Terminal #19618](https://github.com/microsoft/terminal/issues/19618)

### PR DESCRIPTION
**Pull request by gnodet**
## Summary

Add support for [DEC private mode 2048](https://gist.github.com/rockorager/e695fb2924d36b2bcf1fff4a3704bd83) — in-band window resize notifications. Fixes #2009.

### Changes

- **Terminal API** (`Terminal.java`): Added `hasInBandResizeSupport()` and `trackInBandResize(boolean)` following the existing `hasFocusSupport()` / `trackFocus()` pattern
- **DECRQM probing** (`AbstractTerminal.java`): Lazy probe for mode 2048 support using the same DECRQM + DA1 sentinel strategy as mode 2027
- **Input parsing** (`LineReaderImpl.java`): New `terminal-resize` widget bound to `CSI 48 ;` that reads rows/cols/pixel parameters until final byte `t`, updates terminal size, and raises `WINCH`
- **Lifecycle**: Mode is automatically disabled when the terminal is closed
- **Remote terminals**: Mode 2048 works transparently over SSH/Telnet — escape sequences flow through the data stream, so no remote module changes are needed

### Sequences

| Action | Sequence |
|--------|----------|
| Enable | `CSI ? 2048 h` |
| Disable | `CSI ? 2048 l` |
| Query (DECRQM) | `CSI ? 2048 $ p` |
| Report (terminal → app) | `CSI 48 ; rows ; cols ; pixelHeight ; pixelWidth t` |

### Usage

```java
Terminal terminal = TerminalBuilder.terminal();

if (terminal.hasInBandResizeSupport()) {
    terminal.trackInBandResize(true);
    // Resize reports are now delivered in-band as CSI 48 t
    // LineReader handles them automatically via the terminal-resize widget
}
```

## Test plan

- [x] DECRPM probing: set (Ps=1), reset (Ps=2), permanently set (Ps=3) → supported
- [x] DECRPM probing: not recognized (Ps=0), permanently reset (Ps=4) → not supported
- [x] Dumb/dumb-color terminals → not supported, no probe sent
- [x] No DECRPM response (DA1 only) → not supported
- [x] No response at all → not supported
- [x] Probe result is cached
- [x] Enable/disable sends correct escape sequences
- [x] Mode disabled on terminal close
- [x] Default interface methods return false
- [x] All existing terminal tests pass (388 tests)
- [x] All existing reader tests pass (210 tests)
- [x] Full reactor build succeeds

## Summary by CodeRabbit

* **New Features**
  * Added in-band terminal resize reporting for compatible terminals.
  * Added support checks and controls for enabling or disabling resize tracking.
  * Terminal resize events now update dimensions and trigger resize notifications automatically.

* **Bug Fixes**
  * Ensured resize tracking is disabled when a terminal session closes.
  * Improved handling of malformed resize reports.
  * Reduced the default delay for ambiguous key bindings.
  * Improved control-character display and interrupt cleanup behavior.

### CODE REVIEW

**Comment by gnodet** [comment]
Rebased on top of #2023 (batch DEC mode probing). The `hasInBandResizeSupport()` now delegates to `isModeSupported(Mode.IN_BAND_RESIZE)` and the standalone `probeMode2048()` / `probeInBandResize()` methods have been removed — mode 2048 is probed as part of the single batch DECRQM round-trip.

Tests streamlined: low-level DECRPM parsing tests moved to `DecModeProbeTest`; `InBandResizeTest` now focuses on the enable/disable/close lifecycle.

**Comment by gnodet** [comment]
@coderabbitai review

**Review by gnodet — COMMENTED** [review]
**LGTM** — Well-structured implementation of in-band resize notifications (DEC private mode 2048) that follows established project patterns. The API design mirrors the existing `hasFocusSupport()`/`trackFocus()` pair, the batch mode probing is cleanly integrated, and the test coverage for the terminal-level API is thorough.

The `Size.of(cols, rows)` argument order in `applyResizeParams` correctly handles the spec's height-first, width-second reporting format. The `doClose()` ordering correctly disables in-band resize before grapheme cluster mode and before marking the terminal as closed, consistent with the existing pattern. CI is all green across all platforms and Java versions.

_This review was generated by an AI agent and may contain inaccuracies. Please verify all suggestions before applying._

_Claude Code on behalf of Guillaume Nodet_

File: terminal/src/main/java/org/jline/terminal/Terminal.java [hunk 1]
```diff
@@ -1457,6 +1457,58 @@ default boolean isModeSupported(Mode mode) {
...
+     * {@link #isModeSupported(Mode) isModeSupported(Mode.IN_BAND_RESIZE)}.
+     * The probe is never sent to dumb terminals.
+     * </p>
+     *
+     * @return {@code true} if the terminal supports mode 2048
+     * @see #trackInBandResize(boolean)
+     */
+    default boolean hasInBandResizeSupport() {
+        return false;
+    }
```
[the remark this answers is not in the document]
Comment (gnodet) [reply terminal/src/main/java/org/jline/terminal/Terminal.java:1595]: Fixed in 478d742. The default now delegates to `isModeSupported(Mode.IN_BAND_RESIZE)` — both are default methods returning `false`, so the behavior is unchanged for existing implementations, but any custom `Terminal` that overrides `isModeSupported()` will now get correct behavior from the inherited default. Also added `@since 3.30.0` tags to both methods for consistency.

File: reader/src/main/java/org/jline/reader/impl/LineReaderImpl.java [hunk 2]
```diff
@@ -6472,6 +6474,71 @@ public boolean focusOut() {
...
+                sb.append((char) c);
+                if (sb.length() > 50) {
+                    return null; // Too long — discard
+                }
+            } else {
+                return null; // Invalid character — discard
+            }
+        }
+        return null;
+    }
```
[the remark this answers is not in the document]
Comment (gnodet) [reply reader/src/main/java/org/jline/reader/impl/LineReaderImpl.java:6544]: Good catch — fixed in 478d742. `readResizeParams()` now uses a `discard` flag: on invalid character or length > 50 it continues reading until the terminating `t` before returning `null`, preventing leftover bytes from corrupting subsequent key parsing.

File: terminal/src/main/java/org/jline/terminal/Terminal.java [hunk 3]
```diff
@@ -1457,6 +1457,58 @@ default boolean isModeSupported(Mode mode) {
...
+     * parse the reports manually — the reader's built-in
+     * {@code terminal-resize} widget handles them automatically, updating
+     * the terminal size and raising {@link Signal#WINCH}.
+     * </p>
+     *
+     * @param tracking {@code true} to enable in-band resize notifications,
+     *                 {@code false} to disable them
+     * @return {@code true} if the operation succeeded, {@code false} if the
+     *         terminal does not support mode 2048
+     * @see #hasInBandResizeSupport()
```
[the remark this answers is not in the document]
Comment (gnodet) [reply terminal/src/main/java/org/jline/terminal/Terminal.java:1619]: Fixed in 478d742. The `@return` javadoc now explicitly documents the asymmetry: "when enabling, `true` if the terminal supports mode 2048 and the mode was activated, `false` otherwise; when disabling, always `true` (idempotent)." This matches the `AbstractTerminal` implementation where `trackInBandResize(false)` is a no-op that always succeeds.

File: reader/src/main/java/org/jline/reader/impl/LineReaderImpl.java [hunk 4]
```diff
@@ -6472,6 +6474,71 @@ public boolean focusOut() {
...
+     * {@link Terminal.Signal#WINCH}.</p>
+     *
+     * @return {@code true} always
+     */
+    public boolean terminalResize() {
+        String params = readResizeParams();
+        if (params != null) {
+            applyResizeParams(params);
+        }
+        return true;
```
[the remark this answers is not in the document]
Comment (gnodet) [reply reader/src/main/java/org/jline/reader/impl/LineReaderImpl.java:6515]: Added in 478d742. New `InBandResizeWidgetTest` (5 tests) covers:
- Valid resize sequence updates terminal size (`CSI 48;30;100;0;0t` → 100×30)
- Rows+cols only (no pixel dimensions)
- Malformed sequence (invalid character) is drained and discarded without crash
- Too-long sequence (>50 chars) is drained and discarded
- Multiple consecutive resize sequences apply the last one

**Review by gnodet — COMMENTED** [review]
Re-reviewed after new commit `478d742` ("fix: address review feedback on mode 2048 PR").

**Changes in this commit:**

1. **Drain-to-'t' fix** (`readResizeParams()`) — The previous version returned `null` immediately on invalid characters or overlong sequences, leaving orphaned bytes in the input stream that could corrupt subsequent key parsing. The new version uses a `discard` flag to drain all bytes up to and including `t` before returning `null`. This is the right fix — escape sequence parsers should always consume the full sequence.

2. **API contract alignment** — `hasInBandResizeSupport()` default now delegates to `isModeSupported(Mode.IN_BAND_RESIZE)` instead of hardcoding `false`, matching its own javadoc. Since `isModeSupported()` also defaults to `false`, this is backwards-compatible.

3. **Javadoc** — `@since 3.30.0` tags added, `@return` on `trackInBandResize` clarified (enable → may return false; disable → always true).

4. **New tests** — 5 reader-side widget tests covering normal resize, minimal params, malformed input (drain behavior), overlong sequences, and multiple-resize-last-wins.

All changes look correct. LGTM.

_This review was generated by an AI agent and may contain inaccuracies. Please verify all suggestions before applying._
_Claude Code on behalf of Guillaume Nodet_

**Review by gnodet — COMMENTED** [review]
LGTM — solid implementation of DEC private mode 2048 that follows established project patterns cleanly. The API mirrors the existing `hasFocusSupport()`/`trackFocus()` pair, the drain-on-discard fix is correct, `Size.of(cols, rows)` correctly maps the spec's `rows;cols` report format, and test coverage is thorough (5 reader-side + 9 terminal-side tests).

Two minor observations:

1. The default `trackInBandResize(boolean)` in `Terminal.java` returns `false` unconditionally, but the `@return` javadoc says "when disabling, always `true` (idempotent)." A custom Terminal implementation that doesn't override this method would technically violate the documented contract when called with `tracking=false`. In practice this is harmless since `AbstractTerminal` overrides it correctly.

2. The `inBandResizeEnabled` field in `AbstractTerminal` is not `volatile`, unlike the nearby `modesProbed` and `closed` fields. If `trackInBandResize(true)` is called from one thread and `doClose()` from another, there's a theoretical visibility issue. Low risk since terminal usage is typically single-threaded.

_This review was generated by an AI agent and may contain inaccuracies. Please verify all suggestions before applying._

_Claude Code on behalf of Guillaume Nodet_

**Review by gnodet — COMMENTED** [review]
✅ **LGTM** — Clean, well-structured implementation of DEC private mode 2048 (in-band window resize) that follows existing patterns (focus tracking, mode 2027 grapheme cluster) precisely. Comprehensive tests cover the terminal API lifecycle, DECRPM probing, and the LineReader widget including edge cases. No correctness issues found.

Minor observations (informational, non-blocking):
- The `readResizeParams()` method discards sequences containing colons — a safe conservative choice since no colon sub-parameters are currently defined, but future protocol extensions may use them.
- The `@since 3.30.0` annotations are consistent with existing annotations in `Terminal.java` but technically reference a version where these methods won't exist (they'll ship in 4.4.0+).
- The stale Javadoc reference fix (replacing `probeMode2027()` with prose) is correct — that method was removed in the batch probe refactor PR #2023.

### 📋 PR Metadata

| Aspect | Current | Suggested |
|--------|---------|----------|
| Labels | _(none)_ | `enhancement` |
| Milestone | _(none)_ | `4.4.0` |

---
_This review was generated by an AI agent and may contain inaccuracies. Please verify all suggestions before applying._

_Claude Code on behalf of Guillaume Nodet_
